view/pe/teview.cpp

```diff
@@ -227,27 +214,29 @@ bool TEView::Init()
 		ReadTEImageHeader(reader, header);
 		ReadTEImageSectionHeaders(reader, header.numberOfSections);
 		m_headersOffset = header.strippedSize - EFI_TE_IMAGE_HEADER_SIZE;
+		m_imageBase = header.imageBase;
 
 		// Set architecture and platform
-		HandleUserOverrides();
-		if (m_arch)
+		auto settings = GetLoadSettings(GetTypeName());
+		if (settings)
 		{
-			auto archName = m_arch->GetName();
-			if (archName == "x86")
-				platform = Platform::GetByName("efi-x86");
-			if (archName == "x86_64")
-				platform = Platform::GetByName("efi-x86_64");
-			if (archName == "aarch64")
-				platform = Platform::GetByName("efi-aarch64");
-			if (!platform)
+			if (settings->Contains("loader.imageBase"))
 			{
-				m_logger->LogError("TE architecture '%s' is not supported", archName.c_str());
-				return false;
+				uint64_t baseOverride = settings->Get<uint64_t>("loader.imageBase", this);
+				if (baseOverride != header.imageBase)
+				{
+					// TE image bases represent the base of the original PE (before headers are stripped) - make this
+					// adjustment for the user
+					m_imageBase = baseOverride - m_headersOffset;
```

Comment:
There is something strange about this code. We should try to keep the definition of "loader.imageBase" consistent across view types.

Also, the comparison above `baseOverride != header.imageBase` would seem to indicate that "loader.imageBase" and header.imageBase have the same meaning. If they mean the same, and we have this: `m_imageBase = header.imageBase;`, then in the override case m_imageBase should be directly assigned baseOverride.

Reply:
Looking elsewhere in the code, m_image base is used many times like this: `m_imageBase + m_headersOffset`. Perhaps m_imageBase should be header.imageBase - m_headersOffset?

Reply:
This code is strange because it's accounting for the hell that is the TE file format. See [here](https://uefi.org/specs/PI/1.8/V1_TE_Image.html) for more information.

TL;DR TE's are frankenstein'd PEs. The toolchain starts by building a PE file, then it strips off the MS-DOS, COFF, optional headers, and section headers and replaces them with TE headers (which are smaller in size). The top level TE header contains a "strippedSize" member that tells you how much data was stripped from the original PE. `m_imageBase` represents the base of the original PE image (before it was stripped).

From the spec:

> The StrippedSize should be set to the number of bytes removed from the start of the original image, which will typically include the MS-DOS, COFF, and optional headers, as well as the section headers. This size can be used by image loaders and tools to make appropriate adjustments to the other fields in the TE image header. Note that StrippedSize does not take into account the size of the TE image header that will be added to the image. That is to say, the delta in the total image size when converted to TE is StrippedSize - sizeof (EFI_TE_IMAGE_HEADER). This will typically need to be taken into account by tools using the fields in the TE header.

Reply:
For even more context, here are some examples of other loaders that need to make the same adjustments:

https://github.com/mkrasnitski/teloader/blob/master/teloader.py#L68
https://github.com/binarly-io/efiXplorer/blob/1f268587cdd378cd689ae256691434c7a305cc97/efiXloader/pe_ida.h#L1062